[tools] Add clang_format_new.py to download, verify and run clang-format binary

SplitInfinity opened the pull request:
**Summary**
This commit adds `tools/clang_format_new.py`, which downloads a platform-appropriate
clang-format binary to a `.gitignored` location, verifies the binary by comparing its
SHA1 hash to a reference hash (also included in this commit), and runs it on all files
matched a specific regex in a list of whitelisted subdirectories of pytorch.
    
This script will eventually replace `tools/clang_format.py`.
    
**Testing**
Ran the script.

*No Args*
```
pytorch > ./tools/clang_format.py 
Downloading clang-format to /Users/<user>/Desktop/pytorch/.clang-format-bin
0% |################################################################| 100%
Using clang-format located at /Users/<user>/Desktop/pytorch/.clang-format-bin/clang-format
> echo $?
0
> git status
<bunch of files>
```

`--diff` *mode*
```
> ./tools/clang_format.py --diff
Using clang-format located at /Users/<user>/Desktop/pytorch/.clang-format-bin/clang-format
Some files are not formatted correctly
> echo $?
1

<format files using the script>

> ./tools/clang_format.py --diff
Using clang-format located at /Users/<user>/Desktop/pytorch/.clang-format-bin/clang-format
All files are formatted correctly
> echo $?
0
```

Commit:
[tools] Add clang_format_new.py to download, verify and run clang-format binary

Summary:
This commit adds tools/clang_format_new.py, which downloads a platform-appropriate
clang-format binary to a .gitignored location, verifies the binary by comparing its
SHA1 hash to a reference hash (also included in this commit), and runs it on all files
matched a specific regex in a list of whitelisted subdirectories of pytorch.

This script will eventually replace tools/clang_format.py.

Testing:
Ran the script.

suo commented on scripts/apply_clang_format.py:
```diff
@@ -0,0 +1,200 @@
+    # This stack contains all directories that have not been scanned yet.
+    dir_stack = [os.path.join(pytorch_root, target_dir) for target_dir in target_dirs]
+
+    while dir_stack:
```
check out `pathlib`, you can do things like `path.glob(**/*.cpp)` to get all cpp files recursively under `path`. And in general it provides a nice object-oriented API for filesystem access (`foo / bar` for path joining, `foo.exists()` for existence checking, etc.) Probably most of the filesystem manipulation in this script would be cleaner using it.

SplitInfinity replied:
I stole what the old script was doing using `os.walk`, let me know if that's not good enough.

suo replied:
Using `pathlib` would be nice (we previously couldn't because it isn't available in Py2), but it's not that important.

suo commented on scripts/apply_clang_format.py:
```diff
@@ -0,0 +1,200 @@
+    # Make sure the PyTorch root is absolute.
+    pytorch_root = gcf.sanitize_path(options.pytorch_root)
+    # Check that the given PyTorch root is actually the root and change the current working directory to it.
+    ok = validate_pytorch_root(pytorch_root)
```
Since this script is in a known location with respect to the root, might be more robust to just navigate relative to `__file__`, which is an interpreter-set variable corresponding to this file path.

SplitInfinity replied:
Fixed.

suo reviewed (changes requested):
Generally looks good, suggested some python-y things inline. Some additional top-level things:
1. These should probably go in `tools/` not `scripts/`, where the existing `clang-format` and `clang-tidy` stuff lives.
2. We should replace `tools/clang_format.py` with this.
3. This script should be runnable with no arguments and do the right thing (looks like this is already the case, just to keep in mind).

suo commented:
Also, can we include weirder extensions like `.cc` and `.hpp` lol

SplitInfinity commented:
Alright, so the diff is ugly because what I added is, for all intents and purposes, a new file. I realized that the moment this lands, it'll mess with people's pre-commit hooks because there is no longer an option to view and accept changes. The first person to commit something after this lands will also have to disable their pre-commit hook in order to avoid committing all the reformatted files. Should I append them to this PR?

suo reviewed (approved):
lgtm overall! Might want to check in with people who have the git hook still installed to make sure they are not surprised.

SplitInfinity commented:
- Updated reference hash
- Renamed script to `clang_format_new.py` so that I can submit another PR that contains the formatting changes and replaces `clang_format.py`
